File: BFS/boj_10026.py
import sys
from collections import deque

input = sys.stdin.readline

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]


def bfs(x, y):
    q = deque()
    q.append([x, y])
    color = graph[x][y]
    visited[x][y] = 1
    while q:
        x, y = q.popleft()
        for k in range(4):
            nx, ny = x + dx[k], y + dy[k]
            if 0 <= nx < n and 0 <= ny < n:
                if visited[nx][ny] == 0 and graph[nx][ny] == color:
                    visited[nx][ny] = 1
                    q.append([nx, ny])


n = int(input())

graph = [list(input().rstrip()) for _ in range(n)]

visited = [[0] * n for _ in range(n)]

cnt1 = 0
for i in range(n):
    for j in range(n):
        if visited[i][j] == 0:
            bfs(i, j)
            cnt1 += 1

        if graph[i][j] == 'G':
            graph[i][j] = 'R'

visited = [[0] * n for _ in range(n)]
cnt2 = 0
for i in range(n):
    for j in range(n):
        if visited[i][j] == 0:
            bfs(i, j)
            cnt2 += 1

print(cnt1, cnt2)

File: Breadth-first Search/boj_10026.cpp
#include <iostream>
#include <algorithm>
#include <queue>

using namespace std;

char graph[100][100] = {};
int visited[100][100] = {};
int dx[] = { -1, 1, 0, 0 };
int dy[] = { 0, 0, -1, 1 };

void bfs(int x, int y, int n) {
    queue<pair<int, int>> q;
    char color = graph[x][y];

    q.push({ x, y });
    visited[x][y] = 1;

    while (!q.empty()) {
        int cx = q.front().first;
        int cy = q.front().second;
        q.pop();

        for (int k = 0; k < 4; ++k) {
            int nx = cx + dx[k];
            int ny = cy + dy[k];

            if (0 <= nx && nx < n && 0 <= ny && ny < n) {
                if (visited[nx][ny] == 0 && graph[nx][ny] == color) {
                    visited[nx][ny] = 1;
                    q.push({ nx, ny });
                }
            }
        }
    }
}

int main() {
    int n;

    cin >> n;

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cin >> graph[i][j];
        }
    }

    int cnt1 = 0;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            if (visited[i][j] == 0) {
                bfs(i, j, n);
                cnt1++;
            }

            if (graph[i][j] == 'G') {
                graph[i][j] = 'R';
            }
        }
    }

    fill(&visited[0][0], &visited[0][0] + 100 * 100, 0);

    int cnt2 = 0;

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            if (visited[i][j] == 0) {
                bfs(i, j, n);
                cnt2++;
            }
        }
    }

    cout << cnt1 << " " << cnt2 << endl;

    return 0;
}
